File: server/src/services/naverShopping.ts
const NAVER_API_URL = 'https://openapi.naver.com/v1/search/shop.json';

export interface NaverProduct {
  id: string;
  title: string;
  price: number;
  image: string;
  link: string;
  mallName: string;
  brand: string;
  category: string;
}

export interface NaverSearchResult {
  products: NaverProduct[];
  total: number;
  query: string;
}

// HTML 태그 제거
function stripHtml(html: string): string {
  return html.replace(/<[^>]*>/g, '');
}

export async function searchNaverShopping(
  query: string,
  keywords: string[],
  display = 20,
  start = 1,
): Promise<NaverSearchResult> {
  const fetch = (await import('node-fetch')).default;

  // 키워드 조합으로 검색어 최적화
  const searchQuery = query || keywords.slice(0, 3).join(' ');

  const params = new URLSearchParams({
    query: searchQuery,
    display: String(display),
    start: String(start),
    sort: 'sim', // 정확도순
  });

  const res = await fetch(`${NAVER_API_URL}?${params}`, {
    headers: {
      'X-Naver-Client-Id': process.env.NAVER_CLIENT_ID!,
      'X-Naver-Client-Secret': process.env.NAVER_CLIENT_SECRET!,
    },
    signal: AbortSignal.timeout(5000),
  });

  if (!res.ok) {
    if (res.status === 401) {
      throw new Error('네이버 쇼핑 API 인증에 실패했어요. API 키를 확인해 주세요.');
    }
    throw new Error(`상품 검색에 실패했어요 (${res.status})`);
  }

  const data = await res.json() as {
    total: number;
    items: Array<{
      productId: string;
      title: string;
      lprice: string;
      image: string;
      link: string;
      mallName: string;
      brand: string;
      category1: string;
    }>;
  };

  const products: NaverProduct[] = (data.items || []).map((item) => ({
    id: item.productId || String(Math.random()),
    title: stripHtml(item.title),
    price: parseInt(item.lprice, 10) || 0,
    image: item.image,
    link: item.link,
    mallName: item.mallName || '',
    brand: item.brand || '',
    category: item.category1 || '',
  }));

  return {
    products,
    total: data.total || 0,
    query: searchQuery,
  };
}

File: fashion-search/src/search/naver_shopping.py
"""네이버쇼핑 API 단일 아이템 검색."""
import logging
import re

import httpx

from src.config.settings import get_settings

logger = logging.getLogger(__name__)

NAVER_SHOP_URL = "https://openapi.naver.com/v1/search/shop.json"

_HTML_TAG_RE = re.compile(r"<[^>]+>")


def _strip_html(text: str) -> str:
    return _HTML_TAG_RE.sub("", text)


def _normalize_item(raw: dict, category: str) -> dict:
    """네이버쇼핑 응답 → ProductCard 호환 dict."""
    return {
        "product_id": raw.get("productId") or raw.get("mallProductId", ""),
        "title": _strip_html(raw.get("title", "")),
        "price": int(raw.get("lprice") or raw.get("hprice") or 0),
        "image_url": raw.get("image", ""),
        "link": raw.get("link", ""),
        "platform": raw.get("mallName", "네이버쇼핑"),
        "category": category,
        "similarity_score": 0.0,
    }


async def search_items(
    client: httpx.AsyncClient,
    query: str,
    category: str,
    display: int = 50,
    exclude: str = 'used:rental:cbshop',
) -> list[dict]:
    """네이버쇼핑 API 단일 쿼리 검색."""
    settings = get_settings()
    if not settings.naver_client_id or not settings.naver_client_secret:
        logger.warning("[naver_shopping] API 키 미설정 — 빈 결과 반환")
        return []

    try:
        params: dict = {"query": query, "display": display, "sort": "sim"}
        if exclude:
            params["exclude"] = exclude
        response = await client.get(
            NAVER_SHOP_URL,
            params=params,
            headers={
                "X-Naver-Client-Id": settings.naver_client_id,
                "X-Naver-Client-Secret": settings.naver_client_secret,
            },
            timeout=10.0,
        )
        response.raise_for_status()
        items = response.json().get("items", [])
        return [_normalize_item(item, category) for item in items]
    except Exception as exc:
        logger.warning("[naver_shopping] 검색 실패 (query=%s): %s", query, exc)
        return []
